### ISSUE TITLE
rmsprop causing strange loss of accurracy part way through training

### ISSUE THREAD

**Issue by majjoris** [ultralytics/ultralytics#24516]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am currently training CNNs. The chosen base model is YOLOV8 from Ultralytics. The training parameters for the optimizers are the same: 160 epochs, 32 batches, a patience of 30, and an input of 512. However, I noticed strange behavior for rmsprop; it presents a low mAP50-95 compared to other optimizers. The training dataset has 7000 images divided into 11 classes, and the test dataset has around 1200 images.

<img width="489" height="137" alt="Image" src="https://github.com/user-attachments/assets/48f0b3e0-39c1-41f6-b004-26f48862ac99" />

Test results on an RTX 3090 with PyTorch version: 1.13.1+cu116 and CUDA version: 11.6

However, when training using Kaggle with an Nvidia T4 and the same input parameters, the result is completely different.

<img width="493" height="37" alt="Image" src="https://github.com/user-attachments/assets/3ae50e28-8016-4b65-88eb-bb022fdccff5" />

Test results on an Nvidia T4 with PyTorch version: 2.9.0+cu126 and CUDA version: 12.6

Any help and guidance you can provide would be greatly appreciated!
Sorry for my English, I'm Brazilian and I'm using Google Translate.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @majjoris, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly; an Ultralytics engineer will also assist you soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since your report involves differing training behavior across environments and optimizers, if this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In particular, please include the exact training command or script, model used, dataset config, `ultralytics` version, and any logs/plots that reproduce the RMSProp behavior on both systems.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Probably AMP related. You can try training with `amp=False`

**Comment by majjoris** [comment]
Okay, I'll try using `amp=False.`

**Comment by majjoris** [comment]
> Probably AMP related. You can try training with `amp=False`

Thank you very much, it worked, however I had to reduce `lr0` to 0.00001 because of the overshooting.

**Comment by glenn-jocher** [comment]
That points to AMP instability on that environment, and needing a lower `lr0` with `RMSProp` is expected since it can overshoot more easily than `AdamW` or `optimizer=auto`; for context, see the [`check_amp()` reference](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp), and if you want to compare the two setups further, please paste `yolo checks` from both machines.

**Comment by majjoris** [comment]
I ran check_amp and check_yolo in both environments.
<img width="921" height="253" alt="Image" src="https://github.com/user-attachments/assets/d1aedf0a-c614-4acc-8b28-4c7beb5f9c63" /> 
result for the rtx 3090

<img width="797" height="322" alt="Image" src="https://github.com/user-attachments/assets/c2562ded-9070-43d2-b825-ed8cf917021d" />
result for Kaggle

I suspect it might be because of the Torch version and how it manipulates the amp.


> That points to AMP instability on that environment, and needing a lower `lr0` with `RMSProp` is expected since it can overshoot more easily than `AdamW` or `optimizer=auto`; for context, see the [`check_amp()` reference](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp), and if you want to compare the two setups further, please paste `yolo checks` from both machines.

**Comment by glenn-jocher** [comment]
Yes, that’s a reasonable suspicion: [`check_amp()`](https://docs.ultralytics.com/reference/utils/checks#ultralytics.utils.checks.check_amp) explicitly compares FP32 vs AMP outputs, so different `torch`/CUDA stacks can pass on Kaggle and fail on your RTX 3090; I’d keep `amp=False` on that machine or upgrade it to a newer PyTorch/CUDA stack, and if you want us to confirm the environment difference please paste the text output of `yolo checks` or `ultralytics.utils.checks.collect_system_info()` from both setups instead of screenshots.

**Comment by majjoris** [comment]
Sorry, I forgot to mention that I'm using Docker; the image is `cschranz/gpu-jupyter:v1.5_cuda-11.6_ubuntu-20.04`
RTX 3090

```
Ultralytics YOLOv8.0.228 🚀 Python-3.10.10 torch-1.13.1+cu116 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (32 CPUs, 141.6 GB RAM, 801.1/820.0 GB disk)

OS                  Linux-4.15.0-189-generic-x86_64-with-glibc2.31
Environment         Jupyter
Python              3.10.10
Install             pip
RAM                 141.57 GB
CPU                 Intel Xeon E5-2620 v4 2.10GHz
CUDA                11.6

matplotlib          ✅ 3.7.1>=3.3.0
numpy               ✅ 1.23.5>=1.22.2
opencv-python       ✅ 4.8.1.78>=4.6.0
pillow              ✅ 9.4.0>=7.1.2
pyyaml              ✅ 6.0>=5.3.1
requests            ✅ 2.28.2>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 1.13.1+cu116>=1.8.0
torchvision         ✅ 0.14.1+cu116>=0.9.0
tqdm                ✅ 4.65.0>=4.64.0
pandas              ✅ 2.0.0>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
psutil              ✅ 5.9.4
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
```
Kaggle
```
Ultralytics 8.4.51 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
Setup complete ✅ (4 CPUs, 31.4 GB RAM, 6841.9/8062.4 GB disk)

OS                     Linux-6.6.122+-x86_64-with-glibc2.35
Environment            Colab
Python                 3.12.12
Install                pip
Path                   /usr/local/lib/python3.12/dist-packages/ultralytics
RAM                    31.35 GB
Disk                   6841.9/8062.4 GB
CPU                    Intel Xeon CPU @ 2.00GHz
CPU count              4
GPU                    Tesla T4, 14913MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 2.0.2>=1.23.0
matplotlib             ✅ 3.10.0>=3.3.0
opencv-python          ✅ 4.13.0.92>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.4>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.10.0+cu128>=1.8.0
torch                  ✅ 2.10.0+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.25.0+cu128>=0.9.0
psutil                 ✅ 5.9.5>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.19>=2.0.18
{'OS': 'Linux-6.6.122+-x86_64-with-glibc2.35',
 'Environment': 'Colab',
 'Python': '3.12.12',
 'Install': 'pip',
 'Path': '/usr/local/lib/python3.12/dist-packages/ultralytics',
 'RAM': '31.35 GB',
 'Disk': '6841.9/8062.4 GB',
 'CPU': 'Intel Xeon CPU @ 2.00GHz',
 'CPU count': 4,
 'GPU': 'Tesla T4, 14913MiB',
 'GPU count': 2,
 'CUDA': '12.8',
 'Package Info': {'numpy': '✅ 2.0.2>=1.23.0',
  'matplotlib': '✅ 3.10.0>=3.3.0',
  'opencv-python': '✅ 4.13.0.92>=4.6.0',
  'pillow': '✅ 11.3.0>=7.1.2',
  'pyyaml': '✅ 6.0.3>=5.3.1',
  'requests': '✅ 2.32.4>=2.23.0',
  'scipy': '✅ 1.16.3>=1.4.1',
  'torch': '✅ 2.10.0+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.25.0+cu128>=0.9.0',
  'psutil': '✅ 5.9.5>=5.8.0',
  'polars': '✅ 1.35.2>=0.20.0',
  'ultralytics-thop': '✅ 2.0.19>=2.0.18'}}
```

**Comment by glenn-jocher** [comment]
Thanks, that explains it: your RTX 3090 Docker env is on a much older stack (`ultralytics==8.0.228`, `torch==1.13.1+cu116`, CUDA 11.6) than Kaggle (`8.4.51`, `torch==2.10.0+cu128`, CUDA 12.8), so different AMP behavior is expected; to compare fairly, update the container to a current stack, restart the Jupyter kernel after upgrading, and if you stay on that older image then `amp=False` for `RMSProp` is the safest path per the [common issues guide](https://docs.ultralytics.com/guides/yolo-common-issues).

**Comment by majjoris** [comment]
Okay, I'll see what solution I can apply, thank you very much :)

**Comment by glenn-jocher** [comment]
You’re welcome — if it still reproduces after updating the container and restarting Jupyter, please open a new issue with a minimal reproducible example and the text output of `yolo checks` so we can take a closer look.
